### PR TITLE
Added getAllianceID

### PR DESCRIPTION
**Pull request by Omnione**
This addresses the cumbersome way of trying to grab the ID of an alliance or party,
Currently there is no easy way of obtaining these, if im wrong thats fine, i'll remove it.

### CODE REVIEW

**Comment by TeoTwawki** [comment]
While there actually is a way to do this, said way is a pain in the rear and I want to use this instead. My only request is to update the casting style to use the modern C++ stuff instead of old C style cast.
`auto PChar = dynamic_cast<CCharEntity*>(m_PBaseEntity);`

**Comment by Omnione** [comment]
Sounds good to me, ill get on it

**Comment by teschnei** [comment]
if you use it on something that isn't a player, it will crash the server (just check if PChar is not nullptr)

**Comment by Omnione** [comment]
Oh crap your right, one sec ill fix that

**Comment by Omnione** [comment]
K done.

File: src/map/lua/lua_baseentity.cpp [hunk 1]
```diff
@@ -8625,6 +8625,41 @@ inline int32 CLuaBaseEntity::getAllianceSize(lua_State* L)
...
+
+inline int32 CLuaBaseEntity::getAllianceID(lua_State* L)
+{
+    DSP_DEBUG_BREAK_IF(m_PBaseEntity == nullptr);
+
+    uint8 id = 0;
+
+    auto PChar = dynamic_cast<CCharEntity*>(m_PBaseEntity);
+
+    if (PChar != nullptr)
```
Comment (nfaletra) [inline src/map/lua/lua_baseentity.cpp]: Inline this

`if (CCharEntity* PChar = dynamic_cast<CCharEntity*>(m_PBaseEntity))`

File: src/map/lua/lua_baseentity.cpp [hunk 2]
```diff
@@ -8625,6 +8625,41 @@ inline int32 CLuaBaseEntity::getAllianceSize(lua_State* L)
...
+*  Purpose : Returns the id of the alliance/party or player
+*  Example : local allianceId = player:getAllianceID()
+*  Notes   :
+************************************************************************/
+
+inline int32 CLuaBaseEntity::getAllianceID(lua_State* L)
+{
+    DSP_DEBUG_BREAK_IF(m_PBaseEntity == nullptr);
+
+    uint8 id = 0;
```
Comment (nfaletra) [inline src/map/lua/lua_baseentity.cpp]: someone correct me if I'm wrong, but is 0 a valid alliance ID?

File: src/map/lua/lua_baseentity.cpp [hunk 3]
```diff
@@ -8625,6 +8625,41 @@ inline int32 CLuaBaseEntity::getAllianceSize(lua_State* L)
...
+        {
+            id = PChar->PParty->GetPartyID();
+        }
+        else
+        {
+            id = PChar->id;
+        }
+    }
+
+    lua_pushnumber(L, id);
```
Comment (nfaletra) [inline src/map/lua/lua_baseentity.cpp]: I feel this should push nil if you're not in an alliance. Falling back to party ID would not be the behavior I would expect on the script side.

File: src/map/lua/lua_baseentity.cpp [hunk 4]
```diff
@@ -8625,6 +8625,41 @@ inline int32 CLuaBaseEntity::getAllianceSize(lua_State* L)
...
+
+    auto PChar = dynamic_cast<CCharEntity*>(m_PBaseEntity);
+
+    if (PChar != nullptr)
+    {
+        if (PChar->PParty && PChar->PParty->m_PAlliance)
+        {
+            id = PChar->PParty->m_PAlliance->m_AllianceID;
+        }
+        else if (PChar->PParty)
```
Comment (nfaletra) [inline src/map/lua/lua_baseentity.cpp]: Potentially doing the same conditional twice in here.

You can wrap all of this in a PParty check, then check if the character is in an alliance.

If not, you should really be pushing nil. Falling back to party id and then char id is not intuitive behavior.

File: src/map/lua/lua_baseentity.cpp (hunk 1 above)
Comment (nfaletra) [reply src/map/lua/lua_baseentity.cpp]: should also be const.

File: src/map/lua/lua_baseentity.cpp [hunk 5]
```diff
@@ -8625,6 +8625,41 @@ inline int32 CLuaBaseEntity::getAllianceSize(lua_State* L)
...
+        {
+            id = PChar->PParty->GetPartyID();
+        }
+        else
+        {
+            id = PChar->id;
+        }
+    }
```
Comment (TeoTwawki) [inline src/map/lua/lua_baseentity.cpp]: If its never a intended to use this on anything but a player, you should prolly add a debug break
```cpp
DSP_DEBUG_BREAK_IF(m_PBaseEntity->objtype != TYPE_PC);
```
and/or maybe think about changing that luapush and return to only happen a result is attainable?
```cpp
    if (PChar)
    {
        if (PChar->PParty && PChar->PParty->m_PAlliance)
        {
            id = PChar->PParty->m_PAlliance->m_AllianceID;
        }
        else if (PChar->PParty)
        {
            id = PChar->PParty->GetPartyID();
        }
        else
        {
            id = PChar->id;
        }

        lua_pushnumber(L, id);
        return 1;
    }

    return 0;
}
```

Comment (TeoTwawki) [reply src/map/lua/lua_baseentity.cpp]: Just saw nfaletra's comments, was distracted by work and left my edit window up. I think pushing nil makes more sense.

**Comment by Omnione** [comment]
Im starting to think i should split this and have getPartyID and getAllianceID, i only realy did it this way to make my life easier when getting an ID to set to a casket, that way i could do it all with just player:getAllianceID() rather than going through the whole; 
am i in an alliance?
yes: ok do player:getAllianceID()
no: ok then do player:getPartyID()
im not in a party so get my ID

File: src/map/lua/lua_baseentity.cpp (hunk 5 above)
Comment (Omnione) [reply src/map/lua/lua_baseentity.cpp]: Yep I agree, just working through all the suggested changes while answering the door to kids as its halloween.

File: src/map/lua/lua_baseentity.cpp (hunk 1 above)
Comment (Omnione) [reply src/map/lua/lua_baseentity.cpp]: Ok changed to:
`if (const CCharEntity* PChar = dynamic_cast<CCharEntity*>(m_PBaseEntity))`

File: src/map/lua/lua_baseentity.cpp (hunk 2 above)
Comment (Omnione) [reply src/map/lua/lua_baseentity.cpp]: I'm not sure, but i can tell you that i have never seen it myself, but it will push nil now to avoid this.

**Comment by TeoTwawki** [comment]
> Im starting to think i should split this and have getPartyID and getAllianceID, i only realy did it this way to make my life easier when getting an ID to set to a casket, that way i could do it all with just player:getAllianceID() rather than going through the whole;
> am i in an alliance?
> yes: ok do player:getAllianceID()
> no: ok then do player:getPartyID()
> im not in a party so get my ID

I thought falling back to the party made sense, but maybe it would be useful somewhere someday to get a specific party's ID while in an alliance too.

**Comment by Omnione** [comment]
Hopefully that should be ok, if not please let me know, im tired so be gentle lol

File: src/map/lua/lua_baseentity.cpp (hunk 3 above)
Comment (Omnione) [reply src/map/lua/lua_baseentity.cpp]: pushing nil now

File: src/map/lua/lua_baseentity.cpp [hunk 6]
```diff
@@ -8625,6 +8625,41 @@ inline int32 CLuaBaseEntity::getAllianceSize(lua_State* L)
...
+    DSP_DEBUG_BREAK_IF(m_PBaseEntity->objtype != TYPE_PC);
+
+    uint8 id = 0;
+
+    if (const CCharEntity* PChar = dynamic_cast<CCharEntity*>(m_PBaseEntity))
+    {
+        if (PChar->PParty)
+        {
+            if (PChar->PParty->m_PAlliance)
+                id = PChar->PParty->m_PAlliance->m_AllianceID;
```
Comment (nfaletra) [inline src/map/lua/lua_baseentity.cpp]: so this logic should stop and return here. If you have an alliance, push the id and gtfo. If you aren't in an alliance. push nil and move on.

At the very least, I don't necessarily have an issue with returning party ID if alliance doesn't exist, but pushing the character ID is weird as hell. It feels like you're coding this function to suit a specific purpose in a script, but you need to think outside of that use case.

Comment (Omnione) [reply src/map/lua/lua_baseentity.cpp]: Ok no worries, ill take it out

Comment (Omnione) [reply src/map/lua/lua_baseentity.cpp]: you know what, im going to split this into seperate functions, one for party one for alliance

Comment (Omnione) [reply src/map/lua/lua_baseentity.cpp]: Would it be ok to make a specific function to do what i was doing before?, i would like to have a getCasketOwnerID that basicly did what i did in the first place

Comment (Omnione) [reply src/map/lua/lua_baseentity.cpp]: I'll PR that seperately

**Comment by Omnione** [comment]
Ok i split getAllianceId into two now, getAllianceID and getPartyID, nice and simple

File: src/map/lua/lua_baseentity.cpp [hunk 7]
```diff
@@ -8625,6 +8625,61 @@ inline int32 CLuaBaseEntity::getAllianceSize(lua_State* L)
...
+    DSP_DEBUG_BREAK_IF(m_PBaseEntity == nullptr);
+    DSP_DEBUG_BREAK_IF(m_PBaseEntity->objtype != TYPE_PC);
+
+    uint8 id = 0;
+
+    if (const CCharEntity* PChar = dynamic_cast<CCharEntity*>(m_PBaseEntity))
+    {
+        if (PChar->PParty)
+        {
+            id = PChar->PParty->GetPartyID();
```
Comment (TeoTwawki) [inline src/map/lua/lua_baseentity.cpp]: may as well ditch the variable at this point and just move that right into the lua_puchnumber line

Comment (Omnione) [reply src/map/lua/lua_baseentity.cpp]: you know i thought of doing that, wish i did now

Comment (Omnione) [reply src/map/lua/lua_baseentity.cpp]: Done

File: src/map/lua/lua_baseentity.cpp (hunk 6 above)
Comment (teschnei) [reply src/map/lua/lua_baseentity.cpp]: honestly, it doesn't seem that weird to me - an allianceid is just the partyid of the lead party, and a partyid is just the charid of the leader.  at the end of the day, anyone seeking an allianceid is just going to call partyid if its nil and then charid if thats nil too.  (and hopefully nobody is checking if an alliance exists by seeing if it has an id or not)

**Comment by TeoTwawki** [comment]
poor omni is gonna read this and then either instant change it back or stand there confused wondering wtf he should do now hoping for a definitive "do this" answer that may never come.31.10.2019, 20:42, "Tyler Schneider" <notifications@github.com>:@teschnei commented on this pull request.



In src/map/lua/lua_baseentity.cpp:
> +*  Notes   :
+************************************************************************/
+
+inline int32 CLuaBaseEntity::getAllianceID(lua_State* L)
+{
+    DSP_DEBUG_BREAK_IF(m_PBaseEntity == nullptr);
+    DSP_DEBUG_BREAK_IF(m_PBaseEntity->objtype != TYPE_PC);
+
+    uint8 id = 0;
+
+    if (const CCharEntity* PChar = dynamic_cast<CCharEntity*>(m_PBaseEntity))
+    {
+        if (PChar->PParty)
+        {
+            if (PChar->PParty->m_PAlliance)
+                id = PChar->PParty->m_PAlliance->m_AllianceID;

honestly, it doesn't seem that weird to me - an allianceid is just the partyid of the lead party, and a partyid is just the charid of the leader.  at the end of the day, anyone seeking an allianceid is just going to call partyid if its nil and then charid if thats nil too.  (and hopefully nobody is checking if an alliance exists by seeing if it has an id or not)

**Comment by TeoTwawki** [comment]
"email replies do not support markdown" Thats nice github..How about not trashing the emails own plain text formatting along the way at least? Basic readability would be nice..

**Comment by Omnione** [comment]
“Learn from yesterday, live for today, hope for tomorrow.”

Albert Einstein

**Comment by Omnione** [comment]
> an allianceid is just the partyid of the lead party, and a partyid is just the charid of the leader.
This was my reason for doing what i did the first time, and would like it to be this way still,

I understand from a scripting point of view this may be confusing, maybe a change of name to reflect the purpose for this?.

I am sure others will benifit from this for various other reason in the future, or updating previous work for faster functionality or easier readability,

I will leave this for now until i get more feedback, or somone else rewrites it and we go with that,
but lastly i would like to say, if we can keep this as one fallback function i.e alliance->party->id than that would be great but i understand this may not be the best way to handle the situation.

Look forward to everyones feedback.
Omnione(aka MrSent).

**Comment by Omnione** [comment]
oh crap wtf did i do this time >.<

**Pull request merged**